### PR TITLE
usb: rework cdc_acm_poll_out to non-blocking 

### PR DESCRIPTION
**Pull request by jfischer-no**
Change cdc_acm_poll_out to do the best to mimic behavior
of a hardware UART controller without flow control.

With this patch, if the USB subsystem is not ready,
no data is transferred to the buffer, that is, new character
is dropped. If the USB subsystem is ready and the buffer is full,
the first character from the tx_ringbuf is removed to
make room for the new character.

Second patch is to avoid spurious interrupt on configured or resume events.

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]

**Review by eobalski — APPROVED** [review]
lgtm. Can we close #37436 by this?

File: subsys/usb/class/cdc_acm.c [hunk 1]
```diff
@@ -902,16 +897,34 @@ static int cdc_acm_poll_in(const struct device *dev, unsigned char *c)
...
 
-	cdc_acm_fifo_fill(dev, &c, 1);
-	k_sem_take(&dev_data->poll_wait_sem, K_MSEC(100));
+	if (!ring_buf_put(dev_data->tx_ringbuf, &c, 1)) {
+		LOG_INF("Ring buffer full, drain buffer");
+		if (!ring_buf_get(dev_data->tx_ringbuf, NULL, 1)) {
+			LOG_ERR("Failed to drain buffer");
+			return;
+		}
+	}
```
Comment (anangl) [inline subsys/usb/class/cdc_acm.c:925]: Shouldn't there be another call to `ring_buf_put()` done after the room in the buffer is made successfully to actually put the new character there?

**Review by cfriedt — CHANGES_REQUESTED** [review]
see @anangl 's comment

File: subsys/usb/class/cdc_acm.c (hunk 1 above)
Comment (jfischer-no) [reply subsys/usb/class/cdc_acm.c:925]: right, if was first a while, but while could block so I changed it to if but did not think it through, thanks

**Review by anangl — APPROVED** [review]

**Review by cfriedt — APPROVED** [review]

**Pull request merged**